Leian/registrate login

leian7 opened the pull request:
<!--
  This is a guideline for what a PR should look like.
  Feel free to modify it to fit your specific needs.

  Please list the issue this fixes in the PR
-->
Resolves issue #19 

## Changes in this PR.
<!--
  Please include a list of all things this PR will accomplish
  Use the checkbox syntax to show what is done and what is yet to be completed.
  This gives context to the diff.
-->

- [X] Test and implement forms (two: one for User fields and one for UserProfile fields, which includes bonus info such as school and birthdate that don't come bundled with Django's User)
- [X] Test "register" view
- [X] Implement "register" view
- [X] Create register.html template
- [x] Test "login" view
- [x] Implement "login" view
- [x] Fix tests that involve schools

## Testing this PR.
<!--
  Please include a list of explicit instructions for testing this PR.
  If the instructions are 'run `make test`' say that,
  If they are more complicated be thorough.
-->

To run tests:
1. Start up an instance of the StreamWebs docker container 
2. Log into the container w/ ``docker-compose run web bash``
3. cd into /streamwebs_frontend
4. Run ``python manage.py test streamwebs/tests/forms``; ``python manage.py test streamwebs/tests/views``, ``python manage.py test streamwebs/tests/models``

To test registration, login, and logout functionality in-browser:
1. First, either add a school to SCHOOL_CHOICES in /streamwebs_frontend/streamwebs_frontend/settings.py, or uncomment the example provided (line 139). 
2. Run ``docker-compose up``
3. Go to http://localhost:8000/streamwebs/ and follow the links to register, log in, and log out.

### Expected Output.
<!--
  Please insert either a description of what should happen when testing
  your PR or a code-block with terminal output.
-->

All tests should pass. 
Register page looks like: http://i.imgur.com/HQQd2PE.png?1
Log in page looks like: http://i.imgur.com/VlH7JKt.png?1

@Kennric @athai @LyonesGamer

Kennric commented on streamwebs_frontend/streamwebs/tests/forms/test_user_reg_form.py:
```diff
@@ -0,0 +1,24 @@
+    def test_UserForm_fields_exist(self): 
+        user_form = UserForm()
+        for i in range(len(self.expected_fields)):
+            self.assertEqual(user_form.Meta.fields[i], self.expected_fields[i])
```
It may be a good idea to sort self.expected_fields and user_form.Meta.fields - someone could put the right fields in, but in a different order. So you could use this pattern, but use .sorted() on each array - or make each into a set(), then you could do: 
self.assertEqual(set(user_form.Meta.fields), set(self.expected_fields))
without the loop.

Kennric commented on streamwebs_frontend/streamwebs/models.py:
```diff
@@ -1,5 +1,42 @@
+      (SCHOOL_A, 'School A'),
+      (SCHOOL_B, 'School B'),
+      (SCHOOL_C, 'School C'),
+    )
```
I think at some point before this gets merged, you need to abstract the schools out of the tests - maybe have the model get the list of schools from settings or from a schools table, then have your test explicitly over-ride that setting or add a school to the table for the test

Kennric commented:
Please run pep8 or flake8 on your files (pip install flake8, flake8 <filename>) and fix any style issues it points out.

Tests are looking pretty good, I think you are testing what needs to be tested, they just need a little cleanup.

leian7 commented:
@Kennric Thanks, I fixed the field logic using sets and moved the school choices to settings.py. I also fixed both the school and birthdate validators so that they take direct values instead of UserProfile objects, then tweaked their respective tests, and they work great now. But the last test in streamwebs/tests/views/test_registrate_view.py still is not passing; this time I think it is because I'm getting my responses confused. In-browser, I do arrive at the success page after registration.

Kennric commented on streamwebs_frontend/streamwebs/tests/views/test_registrate_view.py:
```diff
@@ -0,0 +1,36 @@
+        profile_form_response = self.client.post(reverse('streamwebs:register'), {'school': ('a'), 'birthdate': '1995-11-10'})
+        self.assertEqual(user_form_response.status_code, 200)
+        self.assertEqual(profile_form_response.status_code, 200)
+        self.assertContains(profile_form_response, 'You have successfully created a StreamWebs account.')
```
you probably want profile_form_response.content to get the actual html of the response.

Or, another thing you could do is test the context - profile_form_response.context will give you access to the variables you passed to the template, so you could test that 'registered' is True. 
https://docs.djangoproject.com/en/1.9/topics/testing/tools/#testing-responses

leian7 replied:
`self.assertTrue(profile_form_response.context['registered'])` actually fails, which is weird because the form itself behaves correctly in-browser

Kennric commented:
looking good - I think you can go a step further with the schools setting in tests. instead of relying on a school 'a' to be set in the settings, your test should explicitly set the schools variable. You can use the python 'with' directive to set a config setting temporarily for a test, 

with self.settings(SCHOOL_CHOICES = ['a','b','c']):
    (your test assertions)

Then, no matter what is in the actual settings file, you can put your test values in for the test.

leian7 commented:
The only major conflict is a migration issue. Once [the user model PR](https://github.com/osuosl/streamwebs/pull/32) is merged into develop, this PR's conflicts can be easily resolved.

athai commented:
:+1: The code looks pretty good to me! Although you may want to squash some of your commits? Here's a guide from docs.o.o (https://docs.osuosl.org/development/git/squashing.html?highlight=squash) if you decide to do so.

Kennric commented:
Looks good to me. +1
